src/Microsoft.AspNet.Routing/Template/TemplateParser.cs

```diff
@@ -134,22 +136,29 @@ private static bool ParseParameter(TemplateParserContext context, TemplateSegmen
 
             while (true)
             {
-                if (context.Current == Separator)
-                {
-                    // This is a dangling open-brace, which is not allowed
-                    context.Error = Resources.TemplateRoute_MismatchedParameter;
-                    return false;
-                }
-                else if (context.Current == OpenBrace)
+                if (context.Current == OpenBrace)
                 {
-                    // If we see a '{' while parsing a parameter name it's invalid. We'll just accept it for now
-                    // and let the validation code for the name find it.
+                    // This is an open brace inside of a parameter, it has to be escaped
+                    if (context.Next())
+                    {
+                        if (context.Current != OpenBrace)
+                        {
+                            context.Error = Resources.TemplateRoute_UnescapedBrace;
+                            return false;
+                        }
+                    }
+                    else
+                    {
+                        // This is a dangling open-brace, which is not allowed
+                        context.Error = Resources.TemplateRoute_MismatchedParameter;
+                        return false;
+                    }
                 }
                 else if (context.Current == CloseBrace)
                 {
+                    // This is a closed brace inside of a parameter. It needs to be escaped, to be valid inside the parameter
```

Comment:
check line-length here. This comment isn't 100% accurate. An unescaped brace is **the end** of the parameter, the comment has the implication that it would be invalid.